intorduce AdaDoRA implementation using both DoRA and AdaLoRa together

OrMullerHahitti opened the pull request:
This pull request introduces support for AdaDoRA (AdaLoRA with DoRA) to the codebase, enabling the use of DoRA (Weight-Decomposed Low-Rank Adaptation) within the AdaLoRA framework. The main changes add new modules and configuration options to support this, update the initialization and management of adapter layers, and provide a new GLUE finetuning script that demonstrates AdaLoRA/DoRA usage.

**Key changes include:**

### AdaDoRA (DoRA for AdaLoRA) Support

* Added the `AdaDoraLinearLayer` class in `adalora/dora.py`, implementing DoRA's magnitude vector and forward logic for AdaLoRA's SVD decomposition, including proper handling of singular values and rank pruning.
* Updated `adalora/__init__.py` to export `AdaDoraLinearLayer` and `AdaDoraLinearVariant` for external use.
* Modified `adalora/config.py` to allow `use_dora` (DoRA) in AdaLoRA by removing the previous error, enabling AdaDoRA configuration.

### Adapter Layer and Variant Management

* Extended `AdaLoraLayer` in `adalora/layer.py` to:
  - Track `lora_magnitude_vector` in `adapter_layer_names` for DoRA support.
  - Add a `lora_variant` dictionary to manage per-adapter variants (e.g., AdaDoRA).
  - Allow initialization of adapter layers with the `use_dora` flag and resolve the correct variant, including initialization of the AdaDoRA variant if requested.

### Utilities and Imports

* Imported `LoraVariant` into `adalora/layer.py` to support adapter variants.

### Example and Usage

* Added a new script `scripts/run_glue.py` that demonstrates finetuning for GLUE tasks with AdaLoRA and AdaDoRA, including argument parsing for DoRA options, correct PEFT configuration, and integration of the new AdaLoRA/DoRA callback.

These changes collectively enable AdaLoRA models to leverage DoRA-style adaptation,while using extending dora.py implementation in LoRA module, with full support in both the core library and example scripts.

BenjaminBossan commented on src/peft/tuners/adalora/dora.py:
```diff
@@ -0,0 +1,192 @@
+        mag_norm_scale = (magnitude / weight_norm).view(1, -1)
+
+        # compute adapter output: x @ (A*E).T @ B.T * scaling / ranknum
+        lora_result = (x @ (lora_A * lora_E).T @ lora_B.T) * (scaling / (ranknum + 1e-5))
```
For LoRA, we took some precautions to make DoRA compatible with FSDP, which requires honoring the forward hooks. Therefore, we don't operate directly on the weights. See:

https://github.com/huggingface/peft/blob/261366de2e40cde64b702d6b9c527081ad850549/src/peft/tuners/lora/dora.py#L71-L74

OrMullerHahitti replied:
Thanks for flagging this, In AdaLoRA, lora_A/lora_B/lora_E are nn.Parameter rather than nn.Linear modules, so we can't use the same forward-hook pattern. We're using `torch.matmul` (@) directly on the parameters.

Do you know if this causes FSDP issues with AdaLoRA? If so, I can look into wrapping them or using a different approach. Let me know if you have a suggestion

BenjaminBossan replied:
Ah, I forgot about the `nn.Parameter` thing, in that case just disregard what I said. This may cause issues with FSDP but it's a bit of an edge case.

Something else: For calculating the `lora_result`, can we not re-use the `lora_weight` that was already computed above?

OrMullerHahitti replied:
Fixed, all shown in commits

BenjaminBossan reviewed (changes requested):
Thanks for opening a new PR to add DoRA support to AdaLoRA. I did a first review pass which focused on the general design of the implementation, I haven't done a detailed check of the DoRA logic or the example yet, as it makes more sense to finalize the integration first. Please check my comments.

I also saw that there are unrelated changes in the PR, like changing capitalization of comments or adding the uv lock file. Could you please revert all those changes? Finally, please run `make style` to ensure the linter is happy.

Moreover, before the PR can be merged, the docs should be updated, but we can leave this for later.

OrMullerHahitti commented on src/peft/tuners/adalora/__init__.py:
```diff
@@ -16,12 +16,23 @@
+from .variants import AdaDoraLinearVariant
+
+
+__all__ = [
```
will take care of. there is really no need from a user view point.

OrMullerHahitti commented on src/peft/tuners/adalora/model.py:
```diff
@@ -299,6 +301,28 @@ def resize_state_dict_by_rank_pattern(self, rank_pattern, state_dict, adapter_na
+                    # get the AdaDoraLinearLayer from lora_magnitude_vector
+                    dora_layer = module.lora_magnitude_vector[adapter_name] if adapter_name in module.lora_magnitude_vector else None
+                    if dora_layer is not None and hasattr(dora_layer, "update_magnitude_after_pruning"):
+                        dora_layer.update_magnitude_after_pruning(
```
Yes, you understood correctly - it's a no-op at the moment. I kept it as an explicit "do nothing" rather than removing it, since resetting the magnitude after pruning actually hurt performance (made the ratio m/||W+ΔW|| = 1, effectively disabling dora).

The only value in keeping them was as documentation ("we explicitly chose not to reset magnitude here").
 a comment in the code or PR description achieves the same thing without the unused code.

Could definitely remove it for cleanliness though - happy to do either

BenjaminBossan commented on src/peft/tuners/adalora/layer.py:
```diff
@@ -76,8 +93,20 @@ def update_layer(
         self._move_adapter_to_device_of_base_layer(adapter_name)
+
+        # initialize variant if needed (e.g., DoRA) - must happen after device move
+        if use_dora:
```
This check can be removed. `if lora_variant is not None` is enough.

BenjaminBossan commented on examples/adalora_with_dora/run_glue.py:
```diff
@@ -0,0 +1,282 @@
+#!/usr/bin/env python
+# Copyright 2020 The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2025 The HuggingFace Inc. team. All rights reserved.
```

OrMullerHahitti commented on src/peft/tuners/adalora/layer.py:
```diff
@@ -255,10 +325,16 @@ def update_ipt(self, model):
                     self.ipt[n] = torch.zeros_like(p)
                     self.exp_avg_ipt[n] = torch.zeros_like(p)
                     self.exp_avg_unc[n] = torch.zeros_like(p)
+                # skip update if no gradient (can happen at first step or for frozen params)
```
I've created two reproducers demonstrating the p.grad is None issue in vanilla AdaLoRA:

`examples/adalora_with_dora/reproducer_grad_none.py`
 - Shows the defensive check prevents crashes when 
update_and_allocate()
 is called before the first backward()

`examples/adalora_with_dora/reproducer_crash_without_check.py`
 - Demonstrates the actual TypeError that occurs without the check
Both can be run with:

```bash
uv run python examples/adalora_with_dora/reproducer_grad_none.py
```
```bash
uv run python examples/adalora_with_dora/reproducer_crash_without_check.py
```
> note: using uv as my package manager

The issue occurs because at the first training step (before backward()), p.grad is None, causing a TypeError on line 347 when computing 
(p * p.grad).abs()

OrMullerHahitti commented:
Summary of Changes
Hi @BenjaminBossan! I've addressed all your review comments and resolved merge conflicts with the latest upstream. Here's a comprehensive summary:

1. Merge Conflict Resolution
Resolved conflicts from upstream's API refactoring that changed update_layer signature:

Old: Individual parameters (lora_dropout, init_lora_weights, use_dora, etc.)
New: Config object (config: AdaLoraConfig)
Solution: Updated AdaDoRA to extract use_dora from config while maintaining full compatibility
Files modified:

``src/peft/tuners/adalora/layer.py``
 - Updated 
update_layer()
 and SVDLinear.__init__() signatures
``src/peft/tuners/adalora/model.py``
 - Updated 
_create_and_replace()
 to pass config object
2. Optimized 
lora_weight
 Computation
Fixed redundant computation in AdaDoraLinearLayer.forward():

Before: Computed lora_weight twice (once for norm, once for result)
After: Compute once, reuse for both operations (detached for norm calculation)
Location: ``src/peft/tuners/adalora/dora.py:118-133``
# compute lora_weight once
lora_weight = self._compute_lora_weight(lora_A, lora_B, lora_E, scaling, ranknum)
# reuse for weight norm (detached)
weight_norm = self.get_weight_norm(weight, lora_weight.detach())
# reuse for adapter output
lora_result = x @ lora_weight.T

3. Fixed lora_magnitude_vector Placement Corrected the location of lora_magnitude_vector initialization:

Before: Incorrectly added to base AdaLoraLayer.adapter_layer_names
After: Dynamically added only by DoRA variant during initialization
Rationale: Vanilla AdaLoRA doesn't use magnitude vectors, only AdaDoRA does
Location: src/peft/tuners/adalora/variants.py:59-61
if not module.lora_magnitude_vector:
    # first dora layer being added, add lora_magnitude_vector to learnable params 
`module.adapter_layer_names` =module.adapter_layer_names[:] + ("lora_magnitude_vector",)
4. Removed Redundant use_dora Check Simplified logic in update_layer():

Before: Checked both if use_dora: and if lora_variant is not None:
After: Only check if lora_variant is not None: 
(sufficient since resolve_lora_variant returns None when use_dora=False)
Location: src/peft/tuners/adalora/layer.py:94-97
5. Enhanced Documentation
Added paper links and citations to README:

Inline links: Added hyperlinks to AdaLoRA and DoRA papers in introduction Citation section: 
Added BibTeX entries for both papers
Location: `examples/adalora_with_dora/README.md`
6. Test Integration
Integrated AdaDoRA into main PEFT test harness:

Added: AdaLoraConfig(use_dora=True) to ALL_CONFIGS test matrix
Deleted: Custom `tests/test_adalora_dora.py` (functionality now covered by parametrized tests)
Coverage: ~30+ test methods across 6 different models
Location: `tests/test_decoder_models.py:86-94`
Test Results: 11/12 tests passing

Vanilla AdaLoRA: 6/6 passing (all models)
AdaDoRA (use_dora=True): 5/6 passing
√ OPT, GPTJ, Llama, Qwen2, Gemma3
x GPT2 (expected - Conv1D incompatibility, not AdaDoRA-specific)
The GPT2 failure is a known limitation with AdaLoRA and Conv1D layers, not introduced by this PR.

BenjaminBossan commented on src/peft/tuners/adalora/__init__.py:
```diff
@@ -21,7 +21,14 @@
+    "RankAllocator",
+    "SVDLinear",
+    "SVDQuantLinear",
+]
```
With the change undone, these should all fit one line again.

BenjaminBossan commented on src/peft/tuners/adalora/layer.py:
```diff
@@ -165,26 +216,41 @@ def get_delta_weight(self, adapter) -> torch.Tensor:
-
-                x = self._cast_input_dtype(x, lora_A.dtype)
-                result += (dropout(x) @ (lora_A * lora_E).T @ lora_B.T) * scaling / ranknum
+            return self.base_layer(x, *args, **kwargs)
```
Why change this?

BenjaminBossan commented on src/peft/tuners/adalora/variants.py:
```diff
@@ -0,0 +1,229 @@
+        lora_B = module.lora_B[active_adapter]
+        lora_E = module.lora_E[active_adapter]
+        scaling = module.scaling[active_adapter]
+        ranknum = module.ranknum[active_adapter].item()
```
```suggestion
        ranknum = module.ranknum[active_adapter]
```
Same

BenjaminBossan commented on tests/test_decoder_models.py:
```diff
@@ -94,6 +94,16 @@
+            "total_step": 1,
+            "use_dora": True,
+        },
+    ),
```
Let's add entries for AdaDoRA to the test matrices of `test_custom_models.py`.
